Add hover effects for `MacosIconButton` and `MacosBackButton` (closes #168)

whiplashoo opened the pull request:
This adds subtle hover effects (background color change) for both widgets, to better match macOS design norms. 

Adds a `hoverColor` property to `MacosIconButton` and `MacosBackButton`. The default hover color values were taken by apps like Notes, Reminders, and Safari, which feature similar icon buttons. In the example screen, you can see the hover effect:


https://user-images.githubusercontent.com/9117427/142167916-28be9764-4914-4eea-b2a9-88c4affb6e65.mov

Finally, while it was not necessary, the default shape of `MacosIconButton` has been changed from `BoxShape.circle` to `BoxShape.rectangle` with `BorderRadius(circular: 7)`. After looking around at other apps in macOS, almost every icon button has a rounded rectangular shape.

## Pre-launch Checklist

- [✅] I have run `dartfmt` on all changed files <!-- THIS IS REQUIRED -->
- [✅] I have incremented the package version as appropriate and updated `CHANGELOG.md` with my changes <!-- THIS IS REQUIRED -->
- [✅] I have added/updated relevant documentation <!-- If relevant -->
- [✅] I have run "optimize/organize imports" on all changed files
- [✅] I have addressed all analyzer warnings as best I could
<!-- - [ ] I have run `flutter pub publish --dry-run` and addressed any warnings --> <!-- MAINTAINER ONLY -->

whiplashoo committed:
add hover effect on MacosIconButton: hoverColor property

whiplashoo committed:
add hover effect on : hoverColor property

whiplashoo committed:
use rectangle shape with border radius as default for MacosIconButton

whiplashoo committed:
show how to disable hover effect with transparent hoverColor

whiplashoo committed:
chore: update package version and changelog

whiplashoo committed:
chore: update documentation on default shape of MacosIconButton

whiplashoo committed:
pubspec.lock update

GroovinChip committed:
Merge branch 'dev' into hover_iconbuttons

GroovinChip commented on example/macos/Podfile.lock:
```diff
@@ -19,4 +19,4 @@ SPEC CHECKSUMS:
 PODFILE CHECKSUM: 6eac6b3292e5142cfc23bdeb71848a40ec51c14c
 
-COCOAPODS: 1.11.2
+COCOAPODS: 1.10.2
```
```suggestion
COCOAPODS: 1.11.2
```

GroovinChip commented on lib/src/buttons/back_button.dart:
```diff
@@ -21,6 +22,9 @@ class MacosBackButton extends StatefulWidget {
   /// The color to fill the space around the icon with.
   final Color? fillColor;
 
+  /// The color of the button's background when the button is hovered (mouse is over the button).
```
```suggestion
  /// The color of the button's background when the mouse hovers over the button.
```

GroovinChip commented on lib/src/buttons/back_button.dart:
```diff
@@ -139,8 +146,27 @@ class MacosBackButtonState extends State<MacosBackButton>
     return MouseRegion(
       cursor: widget.mouseCursor!,
+      onEnter: (e) {
+        setState(() {
+          _isHovered = true;
+        });
```
```suggestion
        setState(() => _isHovered = true);
```

GroovinChip commented on lib/src/buttons/back_button.dart:
```diff
@@ -139,8 +146,27 @@ class MacosBackButtonState extends State<MacosBackButton>
+        });
+      },
+      onExit: (e) {
+        setState(() {
+          _isHovered = false;
+        });
```
```suggestion
        setState(() => _isHovered = false);
```

GroovinChip commented on lib/src/buttons/icon_button.dart:
```diff
@@ -41,6 +42,9 @@ class MacosIconButton extends StatefulWidget {
   /// The color of the button's background when the button is disabled.
   final Color? disabledColor;
 
+  /// The color of the button's background when the button is hovered (mouse is over the button). Set to Colors.transparent to disable the hover effect.
```
```suggestion
  /// The color of the button's background when the mouse hovers over it. 
  /// 
  /// Set to `Colors.transparent` to disable the hover effect.
```

GroovinChip commented on lib/src/buttons/icon_button.dart:
```diff
@@ -55,12 +59,13 @@ class MacosIconButton extends StatefulWidget {
   /// The border radius for the button.
   ///
-  /// This should only be set if setting [shape] to `BoxShape.rectangle`.
+  /// This should only be set if setting [shape] to `BoxShape.rectangle` and defaults
+  /// to `BorderRadius.circular(7.0)`.
```
```suggestion
  /// This should only be set if setting [shape] to `BoxShape.rectangle`. 
  ////
  /// Defaults to `BorderRadius.circular(7.0)`.
```

GroovinChip commented on lib/src/buttons/icon_button.dart:
```diff
@@ -206,6 +216,16 @@ class MacosIconButtonState extends State<MacosIconButton>
     return MouseRegion(
       cursor: widget.mouseCursor!,
+      onEnter: (e) {
+        setState(() {
+          _isHovered = true;
+        });
```
```suggestion
        setState(() => _isHovered = true);
```

GroovinChip commented on lib/src/buttons/icon_button.dart:
```diff
@@ -206,6 +216,16 @@ class MacosIconButtonState extends State<MacosIconButton>
+        });
+      },
+      onExit: (e) {
+        setState(() {
+          _isHovered = false;
+        });
```
```suggestion
        setState(() => _isHovered = false);
```

GroovinChip commented on lib/src/buttons/icon_button.dart:
```diff
@@ -312,6 +340,8 @@ class MacosIconButtonThemeData with Diagnosticable {
   /// The default background color for [MacosIconButton].
   final Color? backgroundColor;
 
+  final Color? hoverColor;
```
```suggestion
  /// The color of the button when the mouse hovers over it.
  final Color? hoverColor;
```

whiplashoo committed:
chore: update docs and formatting

whiplashoo committed:
test: update debugFillProperties test for MacosIconButton to include hoverColor property

whiplashoo committed:
chore: fix unnecessary imports

whiplashoo committed:
Merge branch 'hover_iconbuttons' of https://github.com/whiplashoo/macos_ui into hover_iconbuttons
merge requested

whiplashoo commented:
Updated per the requested changes. I also fixed a failing test for `MacosIconButton`. 

Still failing the GH-pages build with a 403, even though I accepted your invitation before.

GroovinChip commented:
> Still failing the GH-pages build with a 403, even though I accepted your invitation before.

Yes, this is because you need to clone my repo directly, and not use your fork. You can start doing that once we get both your PR's merged. Review coming soon.

GroovinChip reviewed (approved):
LGTM
